File: bots/blair/Util/Client/EmbedParser.js
const Discord = require('discord.js')

module.exports = class EmbedParser {
    constructor (Code) {
        const Content = Code.split('$v')

        var Embed = { fields : [] }, Color = null, Message = '', Buttons = []

        for (var Item of Content) {
            Item = Item.slice(Item.indexOf('{') + 1, Item.lastIndexOf('}'))

            switch (Item) {
                case (Item.startsWith('message:') ? Item : '') : {
                    Message = Item.replace('message:', '').trim()

                    break
                }

                case (Item.startsWith('author:') ? Item : '') : {
                    const Author = Item.replace('author:', '')
                    const Split = Author.split('&&')

                    if (Split[1]) {
                        const Domain = (new URL(Split[1]))

                        if (!Domain || Domain.host !== 'blair.win' && Domain.host !== 'media.discordapp.net' && Domain.host !== 'cdn.discordapp.com') {
                            Split[1] = null
                        }
                    }

                    Embed.author = {
                        name : Split[0].trim(),
                        iconURL : Split[1] ? Split[1].trim() : null,
                        url : Split[2] ? Split[2].trim() : null
                    }

                    break
                }

                case (Item.startsWith('title') ? Item : '') : {
                    Embed.title = Item.replace('title:', '').trim()

                    break
                }

                case (Item.startsWith('url') ? Item : '') : {
                    Embed.url = Item.replace('url:', '').trim() 
                    
                    console.log(Embed.url)

                    break
                }

                case (Item.startsWith('description:') ? Item : '') : {
                    Embed.description = Item.replace('description:', '').trim()

                    break
                }

                case (Item.startsWith('field:') ? Item : '') : {
                    const Field = Item.replace('field:', '')
                    const Split = Field.split('&&')

                    Embed.fields.push({
                        name : Split[0].trim(),
                        value : Split[1].trim().replace('inline', ''),
                        inline : Split[1].trim().includes('inline') ? true : null
                    })

                    break
                }

                case (Item.startsWith('footer:') ? Item : '') : {
                    const Footer = Item.replace('footer:', '')
                    const Split = Footer.split('&&')

                    if (Split[1]) {
                        const Domain = (new URL(Split[1]))

                        if (!Domain || Domain.host !== 'blair.win' && Domain.host !== 'media.discordapp.net' && Domain.host !== 'cdn.discordapp.com') {
                            Split[1] = null
                        }
                    }

                    Embed.footer = {
                        text : Split[0].trim(),
                        iconURL : Split[1] ? Split[1].trim() : null
                    }

                    break
                }

                case (Item.startsWith('image:') ? Item : '') : {
                    var Image = Item.replace('image:', '').trim()

                    const Domain = (new URL(Image)) 

                    if (!Domain || Domain.host !== 'blair.win' && Domain.host !== 'media.discordapp.net' && Domain.host !== 'cdn.discordapp.com') {
                        Image = null
                    }

                    Embed.image = {
                        url : Image
                    }
                }

                case (Item.startsWith('thumbnail:') ? Item : '') : {
                    var Thumbnail = Item.replace('thumbnail:', '').trim()

                    const Domain = (new URL(Thumbnail))

                    if (!Domain || Domain.host !== 'blair.win' && Domain.host !== 'media.discordapp.net' && Domain.host !== 'cdn.discordapp.com') {
                        Thumbnail = null
                    }

                    Embed.thumbnail = {
                        url : Thumbnail
                    }

                    break
                }

                case (Item.startsWith('timestamp') ? Item : '') : {
                    Embed.timestamp = new Date()

                    break
                }

                case (Item.startsWith('color:') ? Item : '') : {
                    Color = Item.replace('color:', '').trim()

                    break
                }

                case (Item.startsWith('button:') ? Item : '') : {
                    const Button = Item.replace('button:', '')
                    const Split = Button.split('&&')
                    
                    Buttons.push(new Discord.ButtonBuilder({
                        label : Split[0].trim(),
                        url : Split[1].trim(),
                        emoji : Split[2] ? Split[2].trim() : null
                    }).setStyle('Link'))
                }
            }
        }

        var Embeds = []
        var Components = []

        if (Embed.author || Embed.title || Embed.url || Embed.description || Embed.fields.length > 0 || Embed.footer || Embed.image || Embed.thumbnail) {
            Embeds = [new Discord.EmbedBuilder(Embed).setColor(Color)]
        }

        if (Buttons.length > 0) {
            Components = [new Discord.ActionRowBuilder().addComponents(Buttons)]
        }

        return {
            content : Message.length > 0 ? Message : null,
            embeds : Embeds,
            components : Components
        }
    }
}

File: abort/utils/embedparser.py
import discord, aiohttp, asyncio, datetime, io, yarl
from discord.ui import Button, View
from io import BytesIO

def get_parts(params):
    params = params.replace("{embed}", "")
    return [p[1:][:-1] for p in params.split("$v")]


async def to_object(params):

    x = {}
    fields = []
    content = None
    timestamp=None
    files = []
    view = discord.ui.View()

    for part in get_parts(params):

        if part.startswith("content:"):
            content = part[len("content:") :]

        if part.startswith("url:"):
            x["url"] = part[len("url:") :]

        if part.startswith("title:"):
            x["title"] = part[len("title:") :]

        if part.startswith("description:"):
            x["description"] = part[len("description:") :]

        if part.startswith("footer:"):
            x["footer"] = part[len("footer:") :]

        if part.startswith("color:"):
            try:
                x["color"] = int(part[len("color:") :].replace(' ', '').replace('#', ''), 16)
            except:
                x["color"] = 0x2F3136

        if part.startswith("image:"):
            x["image"] = {"url": part[len("description:") :]}

        if part.startswith("thumbnail:"):
            x["thumbnail"] = {"url": part[len("thumbnail:") :]}

        if part.startswith("attach:"):
            async with aiohttp.ClientSession() as session:
                async with session.get(part[len("attach:") :].replace(' ', '')) as resp:
                    balls = await resp.read()
            files.append(
                discord.File(io.BytesIO(balls), yarl.URL(part[len("attach:") :].replace(' ', '')).name)
            )

        if part.startswith("author:"):
            z = part[len("author:") :].split(" && ")
            icon_url=None
            url=None
            for p in z[1:]:
                if p.startswith('icon:'):
                    p=p[len('icon:') :]
                    icon_url=p.replace(' ', '')
                elif p.startswith('url:'):
                    p=p[len('url:') :]
                    url=p.replace(' ', '')
            try:
                name = z[0] if z[0] else None
            except:
                name = None

            x["author"] = {"name": name}
            if icon_url:
                x["author"]["icon_url"] = icon_url
            if url:
                x["author"]["url"] = url

        if part.startswith("field:"):
            z = part[len("field:") :].split(" && ")
            value=None
            inline='true'
            for p in z[1:]:
                if p.startswith('value:'):
                    p=p[len('value:') :]
                    value=p
                elif p.startswith('inline:'):
                    p=p[len('inline:') :]
                    inline=p.replace(' ', '')
            try:
                name = z[0] if z[0] else None
            except:
                name = None
            
            if isinstance(inline, str):
                if inline == "true":
                    inline = True

                elif inline == "false":
                    inline = False

            fields.append({"name": name, "value": value, "inline": inline})

        if part.startswith("footer:"):
            z = part[len("footer:") :].split(" && ")
            text=None
            icon_url=None
            for p in z[1:]:
                if p.startswith('icon:'):
                    p=p[len('icon:') :]
                    icon_url=p.replace(' ', '')
            try:
                text = z[0] if z[0] else None
            except:
                pass
                
            x["footer"] = {"text": text}
            if icon_url:
                x["footer"]["icon_url"] = icon_url

        if part.startswith("label:"):
            z = part[len("label:") :].split(" && ")
            label='no label'
            url=None
            for p in z[1:]:
                if p.startswith('link:'):
                    p=p[len('link:') :]
                    url=p.replace(' ', '')
                    
            try:
                label = z[0] if z[0] else None
            except:
                pass
                

            view.add_item(
                discord.ui.Button(
                    style=discord.ButtonStyle.link, label=label, url=url
                )
            )
            
        if part.startswith('image:'):
            z=part[len('image:') :]
            x['image']={'url': z}
            
        if part.startswith('timestamp:'):
            z=part[len('timestamp:') :].replace(' ', '')
            if z == 'true':
                timestamp=True
                
    if not x:
        embed = None
    else:
        x["fields"] = fields
        embed = discord.Embed.from_dict(x)

    if not params.count("{") and not params.count("}"):
        content = params
        
    if timestamp:
        embed.timestamp=datetime.now(__import__('pytz').timezone('America/New_York'))

    return {"content": content, "embed": embed, "files": files, "view": view}